include/rocksdb/file_system.h

```diff
@@ -0,0 +1,1353 @@
+  enum class IOPriority : uint8_t {
+    kIOLow,
+    kIOHigh,
+    kIOTotal,
+  };
+
+  // Type of the data begin read/written. It can be passed down as a flag
+  // for the FileSystem implementation to optionally handle different types in
+  // different ways
+  enum class IOType : uint8_t {
+    kData,
+    kFilter,
+    kIndex,
+    kMetadata,
+    kWAL,
+    kManifest,
+    kLog,
+    kUnknown,
+    kInvalid,
+  };
+
+  // Per-request options that can be passed down to the FileSystem
+  // implementation. These are hints and are not necessarily guaranteed to be
+  // honored. More hints can be added here in the future to indicate things like
+  // storage media (HDD/SSD) to be used, replication level etc.
+  struct IOOptions {
+    // Timeout for the operation in milliseconds
+    std::chrono::milliseconds timeout;
+
+    // Priority - high or low
+    IOPriority prio;
+
+    // Type of data being read/written
+    IOType type;
+  };
+
+  // A structure to pass back some debugging information from the FileSystem
+  // implementation to RocksDB in case of an IO error
+  struct IODebugContext {
+    // file_path to be filled in by RocksDB in case of an error
+    std::string file_path;
+
+    // A map of counter names to values - set by the FileSystem implementation
+    std::map<std::string, uint64_t> counters;
+
+    // To be set by the FileSystem implementation
+    std::string msg;
+
+    IODebugContext() {}
+
+    void AddCounter(std::string& name, uint64_t value) {
+      counters.emplace(name, value);
+    }
+
+    std::string ToString() {
+      std::ostringstream ss;
+      ss << file_path << ", ";
+      for (auto counter : counters) {
+        ss << counter.first << " = " << counter.second << ",";
+      }
+      ss << msg;
+      return ss.str();
+    }
+  };
+
+  FileSystem();
+
+  virtual ~FileSystem();
+
+  static const char* Type() { return "FileSystem"; }
+
+  // Loads the environment specified by the input value into the result
+  static Status Load(const std::string& value,
+                     std::shared_ptr<FileSystem>* result);
+
+  // Return a default fie_system suitable for the current operating
+  // system.  Sophisticated users may wish to provide their own Env
+  // implementation instead of relying on this default file_system
+  //
+  // The result of Default() belongs to rocksdb and must never be deleted.
+  static std::shared_ptr<FileSystem> Default();
+
+  // Create a brand new sequentially-readable file with the specified name.
+  // On success, stores a pointer to the new file in *result and returns OK.
+  // On failure stores nullptr in *result and returns non-OK.  If the file does
+  // not exist, returns a non-OK status.
+  //
+  // The returned file will only be accessed by one thread at a time.
+  virtual IOStatus NewSequentialFile(const std::string& fname,
+                                     const IOOptions& io_opts,
+                                     const EnvOptions& env_opts,
```

Comment:
Do we need to put EnvOptions here? Ideally we don't. I know consolidating the two options may be painful, but fow now, can we build a two way converter so that the API can be clean?

Reply:
The scope of EnvOptions is a file, for the duration of the file being open. The scope of IOOptions is the specific IOs that would happen as part of file open. How about having a FileOptions with IOOptions embedded in it? Other methods such as FSRandomAccessFile::Read() can take IOOptions as an argument, while methods such as FieSystem::NewRandomAccessFile() can take a FileOptions as argument.